metricq_source_http/source.py

```diff
@@ -0,0 +1,441 @@
+        assert all(isinstance(host, str) for host in hosts)
+        return hosts
+
+    @classmethod
+    def _create_from_host_config(
+        cls,
+        source: "HttpSource",
+        *,
+        hosts: str | list[str],
+        names: str | list[str],
+        **kwargs: Any,
+    ) -> Iterable["Host"]:
+        hosts = cls._parse_hosts(hosts)
+        names = cls._parse_hosts(names)
+        if len(hosts) != len(names):
+            raise ConfigError("Real of names and hosts differ")
+        for host, name in zip(hosts, names):
+            yield Host(source=source, host=host, name=name, **kwargs)
+
+    @classmethod
+    def create_from_host_configs(
+        cls, source: "HttpSource", host_configs: Sequence[dict[str, Any]]
+    ) -> Iterable["Host"]:
+        for host_config in host_configs:
+            yield from cls._create_from_host_config(source, **host_config)
+
+    @property
+    def metadata(self) -> dict[str, metricq.MetadataDict]:
+        return {
+            metric: metadata
+            for group in self._groups.values()
+            for metric, metadata in group.metadata.items()
+        }
+
+    async def task(self, stop_future: asyncio.Future[None]) -> None:
+        async with aiohttp.ClientSession(
+            base_url=URL.build(scheme=self._scheme, host=self._host),
+            timeout=aiohttp.ClientTimeout(total=self.source.http_timeout),
+            **self._authorization_manager.session_params,
+        ) as session:
+            await self._authorization_manager.authorize_session(session)
+            await asyncio.gather(
+                *[
+                    group.task(stop_future, self._authorization_manager, session)
+                    for group in self._groups.values()
+                ]
+            )
+
+
+class HttpSource(metricq.Source):
+    default_interval: Optional[int | float] = None
+    hosts: Optional[list[Host]] = None
+    _host_task_stop_future: Optional[asyncio.Future[None]] = None
+    _host_task: Optional[asyncio.Task[None]] = None
+
+    @metricq.rpc_handler("config")
+    async def _on_config(
+        self,
+        *,
+        hosts: list[dict[str, Any]],
+        http_timeout: int | float = 5,
+        **config: Any,
+    ) -> None:
+        self.default_interval = _extract_interval(**config)
+        self.http_timeout = http_timeout
+
+        if self.hosts is not None:
+            await self._stop_host_tasks()
+
+        self.hosts = list(Host.create_from_host_configs(self, hosts))
+
+        await self.declare_metrics(
+            {
+                metric: metadata
+                for host in self.hosts
+                for metric, metadata in host.metadata.items()
+            }
+        )
+
+        self._create_host_tasks()
+
+    async def _stop_host_tasks(self) -> None:
+        assert self._host_task_stop_future is not None
+        assert self._host_task is not None
+        self._host_task_stop_future.set_result(None)
+        # TODO add timeout
```

Comment:
Timeout

Reply:
Implemented in https://github.com/metricq/metricq-source-http/compare/3f4bb1bfd872...443b9e67a48d#diff-42c22b6a7860bf93b087648cb5623a85d8fb88674eb86d8d00b63f947e3b0157R420-R424